// firebase-tools/lib/init/features/frameworks/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createBackend = exports.getOrCreateBackend = exports.doSetup = void 0;
const clc = require("colorette");
const utils = require("../../../utils");
const logger_1 = require("../../../logger");
const prompt_1 = require("../../../prompt");
const constants_1 = require("./constants");
const repo = require("./repo");
const poller = require("../../../operation-poller");
const api_1 = require("../../../api");
const gcp = require("../../../gcp/frameworks");
const frameworks_1 = require("../../../gcp/frameworks");
const error_1 = require("../../../error");
const frameworksPollerOptions = {
    apiOrigin: api_1.frameworksOrigin,
    apiVersion: frameworks_1.API_VERSION,
    masterTimeout: 25 * 60 * 1000,
    maxBackoff: 10000,
};
async function doSetup(setup, projectId) {
    setup.frameworks = {};
    utils.logBullet("First we need a few details to create your service.");
    await (0, prompt_1.promptOnce)({
        name: "serviceName",
        type: "input",
        default: "acme-inc-web",
        message: "Create a name for your service [1-30 characters]",
    }, setup.frameworks);
    await (0, prompt_1.promptOnce)({
        name: "region",
        type: "list",
        default: constants_1.DEFAULT_REGION,
        message: "Please select a region " +
            `(${clc.yellow("info")}: Your region determines where your backend is located):\n`,
        choices: constants_1.ALLOWED_REGIONS,
    }, setup.frameworks);
    utils.logSuccess(`Region set to ${setup.frameworks.region}.`);
    logger_1.logger.info(clc.bold(`\n${clc.white("===")} Deploy Setup`));
    await (0, prompt_1.promptOnce)({
        name: "deployMethod",
        type: "list",
        default: constants_1.DEFAULT_DEPLOY_METHOD,
        message: "How do you want to deploy",
        choices: constants_1.ALLOWED_DEPLOY_METHODS,
    }, setup.frameworks);
    const backend = await getOrCreateBackend(projectId, setup);
    if (backend) {
        utils.logSuccess(`Successfully created a backend: ${backend.name}`);
    }
}
exports.doSetup = doSetup;
function toBackend(cloudBuildConnRepo) {
    return {
        codebase: {
            repository: `${cloudBuildConnRepo.name}`,
            rootDirectory: "/",
        },
        labels: {},
    };
}
async function getOrCreateBackend(projectId, setup) {
    const location = setup.frameworks.region;
    const deployMethod = setup.frameworks.deployMethod;
    try {
        return await getExistingBackend(projectId, setup, location);
    }
    catch (err) {
        if (err.status === 404) {
            logger_1.logger.info("Creating new backend.");
            if (deployMethod === "github") {
                const cloudBuildConnRepo = await repo.linkGitHubRepository(projectId, location);
                const backendDetails = toBackend(cloudBuildConnRepo);
                return await createBackend(projectId, location, backendDetails, setup.frameworks.serviceName);
            }
        }
        else {
            throw new error_1.FirebaseError(`Failed to get or create a backend using the given initialization details: ${err}`);
        }
    }
    return undefined;
}
exports.getOrCreateBackend = getOrCreateBackend;
async function getExistingBackend(projectId, setup, location) {
    let backend = await gcp.getBackend(projectId, location, setup.frameworks.serviceName);
    while (backend) {
        setup.frameworks.serviceName = undefined;
        await (0, prompt_1.promptOnce)({
            name: "existingBackend",
            type: "confirm",
            default: true,
            message: "A backend already exists for the given serviceName, do you want to use existing backend? (yes/no)",
        }, setup.frameworks);
        if (setup.frameworks.existingBackend) {
            logger_1.logger.info("Using the existing backend.");
            return backend;
        }
        await (0, prompt_1.promptOnce)({
            name: "serviceName",
            type: "input",
            default: "acme-inc-web",
            message: "Please enter a new service name [1-30 characters]",
        }, setup.frameworks);
        backend = await gcp.getBackend(projectId, location, setup.frameworks.serviceName);
        setup.frameworks.existingBackend = undefined;
    }
    return backend;
}
async function createBackend(projectId, location, backendReqBoby, backendId) {
    const op = await gcp.createBackend(projectId, location, backendReqBoby, backendId);
    const backend = await poller.pollOperation(Object.assign(Object.assign({}, frameworksPollerOptions), { pollerName: `create-${projectId}-${location}-${backendId}`, operationResourceName: op.name }));
    return backend;
}
exports.createBackend = createBackend;
